compiler: Fix identifier normalization function

hunger opened the pull request:
`__1` is a valid identifier, which we normalized to `--1`, which is invalid.

This changes the normalization function to leave a `_` in the first position.

This is probably a bit slower than what we had before:-/

ogoffart reviewed (approved):
Thanks.  (Great use of smol_str::SmolStrBuilder)

Could you add a `out` property that starts with `__` in tests/cases/properties/dashes.slint , and maybe add an element with an id that starts with `__` and even a struct or enum that starts with that, to be sure that lookup works and stuff. (I don't believe it would break, but I'd rather check)

hunger commented:
New and improved: More tests ;-)

ogoffart commented on tests/cases/properties/dashes.slint:
```diff
@@ -22,19 +31,25 @@ const Test_Case &instance = *handle;
 assert_eq(instance.get_property_x1(), 42);
 assert_eq(instance.get_property_x2(), 42);
 assert_eq(instance.get_this__has_6_slashes__(), 86);
+assert_eq(instance.get___test_property_, 42);
+assert_eq(instance.get___test_property2, 23);
```
```suggestion
assert_eq(instance.get___test_property_(), 42);
assert_eq(instance.get___test_property2(), 23);
```

ogoffart reviewed (commented):
Thanks for adding the tests.
The faillure in the C++ test is a simple typo.

But the faillure in the node test shows an actual problem:
There are (two) other `normalize_identifier` function in interpreter/api.rs that needs to updated/removed as well.

hunger commented:
This PR is growing on me :-)

* Fixed the nodejs code to translate the names of `global`s to JS syntax
* Deduplicated normalization code from the interpreter
* Used that normalization code over hand-rolled normalization code
* Changed an internal BTreeMap from String to SmolStr keys (to avoid converting back from SmolStr to String)
